app/javascript/src/ChangePassword/components/ChangePassword.jsx

```diff
@@ -18,25 +18,109 @@ type Props = {
   errors: (?FlashMessage)[]
 }
 
+const PASSWORD = 'user[password]'
+const PASSWORD_CONFIRMATION = 'user[password_confirmation]'
+
 const validationConstraints = {
-  'user[password]': {
+  [PASSWORD]: {
     presence: true,
     length: { minimum: 1 }
   },
-  'user[password_confirmation]': {
+  [PASSWORD_CONFIRMATION]: {
     presence: true,
-    length: { minimum: 1 }
+    length: { minimum: 1 },
+    equality: PASSWORD
   }
 }
 
-const ChangePassword = ({ lostPasswordToken, url, errors }: Props) => {
-  const [isFormDisabled, setIsFormDisabled] = useState(true)
+const isValidForm = (form) => validateForm(form, validationConstraints)
+
+const useFormState = () => {
+  const [isDisabled, setIsDisabled] = useState(true)
 
   const onChange = (event) => {
-    const errors = validateForm(event.currentTarget, validationConstraints)
-    setIsFormDisabled(!!errors)
+    const errors = isValidForm(event.currentTarget)
+    setIsDisabled(!!errors)
+  }
+
+  return {
+    isDisabled,
+    onChange
+  }
+}
+
+const usePasswordState = () => {
+  const [value, setValue] = useState('')
+  const [isValid, setIsValid] = useState(undefined)
+
+  const onChange = (value, event) => {
+    setValue(value)
+    setIsValid(!!validateSingleField(event))
+  }
+
+  const onBlur = (event) => {
+    const errors = validateSingleField(event)
+    setIsValid(!!errors)
+  }
+
+  return {
+    value,
+    isValid,
+    onChange,
+    onBlur
+  }
+}
+
+const getPasswordConfirmationErrors = (event) => {
+  const errors = isValidForm(event.currentTarget.parentNode.parentNode)
+  const passwordConfirmationError = (errors && errors[PASSWORD_CONFIRMATION]) && errors[PASSWORD_CONFIRMATION][0]
+  const passwordsMatchError = passwordConfirmationError && passwordConfirmationError.includes('is not equal to')
+  return {
+    passwordConfirmationError,
+    passwordsMatchError
+  }
+}
+
+const shouldInitValidation = (value, event) => {
+  const passwordLength = event.currentTarget.parentNode.parentNode[PASSWORD].value.length
+  return value.length >= passwordLength
+}
+
+const usePasswordConfirmationState = () => {
+  const [value, setValue] = useState('')
+  const [isValid, setIsValid] = useState(undefined)
+  const [passwordDoesntMatch, setPasswordDoesntMatch] = useState(undefined)
+  const [startValidating, setStartValidating] = useState(false)
+
+  const onChange = (value, event) => {
+    setValue(value)
+    const { passwordConfirmationError, passwordsMatchError } = getPasswordConfirmationErrors(event)
+    setPasswordDoesntMatch(passwordsMatchError)
+    const initValidation = shouldInitValidation(value, event)
+    const validationStarted = startValidating || isValid !== undefined
```

Comment:
There's quite a lot of logic being stored in order to know if the form needs to be validated or not. I'd say it's better to do validation when the form is not pristine, or not. That kind of check asks for a simpler way to do it

Reply:
Well, I agree there is a lot of logic here. But the point is I was explicitly asked (@thomasmaas @cntlsn ) to implement validation not only when pristine/not pristine, but according to several scenarios like:
User tabs and input looses focus (empty)
User tabs and input looses input (filled)
Password confirmation validation should start only after it's length is ` >=` to password length
After initiating validation it shouldn't stop (no matter the length)

Reply:
Maybe extract this hook and make it more generic (take 2 password forms, not a Node). This way we can simplify the logic and make it easier to follow.

Reply:
I'm passing the whole form (formNode) because validate.js is doing the validation of all the inputs, including the equality check. If we pass the 2 inputs separated, we'd need to implement that equality check by ourselves, and the library would be useless.

Reply:
We shouldn't worry about `validate.js`, it's only a tool and a disposable one. We should try to craft out code the most independent / generic the better. In any case, I think the hook should be defined elsewhere so that it can be better tested and reviewd.